File: libs/moqtail-ts/src/model/control/requests_blocked.ts
import { ByteBuffer, FrozenByteBuffer, BaseByteBuffer } from '../common/byte_buffer'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'

export class RequestsBlocked {
  public readonly maximumRequestId: bigint

  constructor(maximumRequestId: bigint | number) {
    this.maximumRequestId = BigInt(maximumRequestId)
  }

  getType(): ControlMessageType {
    return ControlMessageType.RequestsBlocked
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.RequestsBlocked)
    const payload = new ByteBuffer()
    payload.putVI(this.maximumRequestId)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('RequestsBlocked::serialize', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): RequestsBlocked {
    const maximumRequestId = buf.getVI()
    return new RequestsBlocked(maximumRequestId)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('RequestsBlocked', () => {
    test('roundtrip', () => {
      const maximumRequestId = 12345n
      const requestsBlocked = new RequestsBlocked(maximumRequestId)
      const frozen = requestsBlocked.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.RequestsBlocked))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = RequestsBlocked.parsePayload(frozen)
      expect(deserialized.maximumRequestId).toBe(maximumRequestId)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const maximumRequestId = 12345n
      const requestsBlocked = new RequestsBlocked(maximumRequestId)
      const serialized = requestsBlocked.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.RequestsBlocked))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = RequestsBlocked.parsePayload(frozen)
      expect(deserialized.maximumRequestId).toBe(maximumRequestId)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const maximumRequestId = 12345n
      const requestsBlocked = new RequestsBlocked(maximumRequestId)
      const serialized = requestsBlocked.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        RequestsBlocked.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/requests_blocked.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::control_message::ControlMessageTrait;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::control::constant::ControlMessageType;
use crate::model::error::ParseError;
use bytes::{BufMut, Bytes, BytesMut};

#[derive(Debug, Clone, PartialEq)]
pub struct RequestsBlocked {
  pub maximum_request_id: u64,
}

impl ControlMessageTrait for RequestsBlocked {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::RequestsBlocked)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.maximum_request_id)?;

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "RequestsBlocked::serialize",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let maximum_request_id = payload.get_vi()?;
    Ok(Box::new(RequestsBlocked { maximum_request_id }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::RequestsBlocked
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let maximum_request_id = 12345;
    let requests_blocked = RequestsBlocked { maximum_request_id };
    let mut buf = requests_blocked.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::RequestsBlocked as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = RequestsBlocked::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, requests_blocked);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let maximum_request_id = 12345;
    let requests_blocked = RequestsBlocked { maximum_request_id };

    let serialized = requests_blocked.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::RequestsBlocked as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = RequestsBlocked::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, requests_blocked);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let maximum_request_id = 12345;
    let requests_blocked = RequestsBlocked { maximum_request_id };
    let mut buf = requests_blocked.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::RequestsBlocked as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = RequestsBlocked::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
